primitives/src/reed_solomon_code/mod.rs

```diff
@@ -131,23 +136,51 @@ where
     Ok(f)
 }
 
+/// Like [`reed_solomon_erasure_decode`] except input points are drawn from the
+/// given FFT domain.
+///
+/// Differences from [`reed_solomon_erasure_decode`]:
+/// - First part of the share is an index into `domain`
+pub fn reed_solomon_erasure_decode_rou<F, D>(
+    shares: D,
+    data_size: usize,
+    domain: &Radix2EvaluationDomain<F>,
+) -> Result<Vec<F>, PrimitivesError>
+where
+    F: FftField,
+    D: IntoIterator,
+    D::Item: Borrow<(usize, F)>,
+    D::IntoIter: ExactSizeIterator + Clone,
+{
+    let domain_shares = shares.into_iter().map(|share| {
+        let &(index, eval) = share.borrow();
+        // TODO(Gus) nth runtime is linear in index!
+        (domain.elements().nth(index).unwrap(), eval)
```

Comment:
We might first collect all elements and then access them here?

Reply:
I thought of that, and it might be the least bad solution. My initial worry is that most elements will never be accessed, so you'll allocate and compute a bunch of elements that are never used. eg. if the code rate is 1/4 then only 25% of elements will be accessed.

## Other ideas

- Another stupid idea is to sort the shares by index so that only a single pass through `elements` is needed, but sorting is ridiculous and you still compute all the elements.
- Another idea is to compute `nth` via log-time scalar multiplication. Still not great but perhaps better than scanning through all elements.

Reply:
Actually the scaler mall solution is n log n as opposed to oven if we collect. No good answers here

Reply:
I prefer the collection idea over n log n for the following reason:
- Code is cleaner and more understandable, and you don't have to touch the low level details of the underlying `EvaluationDomain`.
- Efficiency isn't much an issue here. It's only going to be called once during the decoding phase, and (even considering the memory allocation) the overhead here is insignificant compared to other part.
- The collected evaluation points might be useful in the faster RS decoding.

Reply:
agreed. fixed in https://github.com/EspressoSystems/jellyfish/pull/309/commits/08e870f3ac01fa5fa6bf0f3a3be1d1de7f369b6a